sbi/neural_nets/density_estimators/base.py

```diff
@@ -1,30 +1,54 @@
+from typing import Optional, Tuple
+
 import torch
 from torch import Tensor, nn
 
 
-class DensityEstimator:
+class DensityEstimator(nn.Module):
     r"""Base class for density estimators.
 
     The density estimator class is a wrapper around neural networks that
     allows to evaluate the `log_prob`, `sample`, and provide the `loss` of $theta,x$
     pairs.
+
+    Note:
+        We assume that the input to the density estimator is a tensor of shape
+        (batch_size, d), where d is the dimensionality of the input. The condition
+        is a tensor of shape (batch_size, *condition_shape), where condition_shape
+        is the shape of the condition tensor.
+
     """
 
-    def __init__(self, net: nn.Module) -> None:
+    def __init__(self, net: nn.Module, condition_shape: torch.Size) -> None:
         r"""Base class for density estimators.
 
         Args:
             net: Neural network.
+            condition_shape: Shape of the input. If not provided, it will assume a 1D
+                             input.
         """
-
+        super().__init__()
         self.net = net
+        self._condition_shape = condition_shape
 
     def log_prob(self, input: Tensor, condition: Tensor, **kwargs) -> Tensor:
-        r"""Return the batched log probabilities of the inputs given the conditions.
+        r"""Return the log probabilities of the inputs given a condition or multiple
+        i.e. batched conditions.
+
+        Note:
+            This function should support PyTorch's automatic broadcasting. This means
+            the function should behave as follows for different input and condition
+            shapes:
+            - (d,) + (b,*condition_shape) -> (b,)
```

Comment:
I do not understand the first three cases, why is the output always the same? 
e.g., for the first case, given `d` thetas, and `b` conditions, shouldn't I get `d*b` outputs, in some shape?

Reply:
it would be good to introduce the notions of `d` and `b` etc somewhere, e.g., below in the Args descriptions.

Reply:
Ahh, yeah I see. WIth "d" i meant the dimension of the input without any batch dimension. So if you log_prob it (d,) will always vanish.